pallets/apollo-platform/src/lib.rs

```diff
@@ -0,0 +1,1429 @@
+
+                // Add pool to PoolsByBlock map
+                let num_of_pools = <PoolsByBlock<T>>::iter().count() as u32;
+                let block_number: BlockNumberFor<T> = num_of_pools.into();
+                <PoolsByBlock<T>>::insert(block_number, asset_id);
+            }
+
+            Self::deposit_event(Event::PoolAdded(user, asset_id));
+            Ok(().into())
+        }
+
+        /// Lend token
+        #[pallet::call_index(1)]
+        #[pallet::weight(<T as Config>::WeightInfo::lend())]
+        pub fn lend(
+            origin: OriginFor<T>,
+            lending_asset: AssetIdOf<T>,
+            lending_amount: Balance,
+        ) -> DispatchResultWithPostInfo {
+            let user = ensure_signed(origin)?;
+
+            ensure!(lending_amount > 0, Error::<T>::InvalidLendingAmount);
+            let mut pool_info =
+                <PoolData<T>>::get(lending_asset).ok_or(Error::<T>::PoolDoesNotExist)?;
+            ensure!(!pool_info.is_removed, Error::<T>::PoolIsRemoved);
+
+            // Add lending amount and interest to user if exists, otherwise create new user
+            if let Some(mut user_info) = <UserLendingInfo<T>>::get(lending_asset, user.clone()) {
+                // Calculate interest in APOLLO token
+                let block_number = <frame_system::Pallet<T>>::block_number();
+                let interests =
+                    Self::calculate_lending_earnings(&user_info, &pool_info, block_number);
+                user_info.lending_interest += interests.0 + interests.1;
+                user_info.lending_amount += lending_amount;
+                user_info.last_lending_block = <frame_system::Pallet<T>>::block_number();
+                <UserLendingInfo<T>>::insert(lending_asset, user.clone(), user_info);
+            } else {
+                let new_user_info = LendingPosition {
+                    lending_amount,
+                    lending_interest: 0,
+                    last_lending_block: <frame_system::Pallet<T>>::block_number(),
+                };
+                <UserLendingInfo<T>>::insert(lending_asset, user.clone(), new_user_info);
+            }
+
+            // Transfer lending amount to pallet
+            Assets::<T>::transfer_from(&lending_asset, &user, &Self::account_id(), lending_amount)
+                .map_err(|_| Error::<T>::CanNotTransferLendingAmount)?;
+            pool_info.total_liquidity += lending_amount;
+            <PoolData<T>>::insert(lending_asset, pool_info);
+
+            Self::deposit_event(Event::Lended(user, lending_asset, lending_amount));
+            Ok(().into())
+        }
+
+        /// Borrow token
+        #[pallet::call_index(2)]
+        #[pallet::weight(<T as Config>::WeightInfo::borrow())]
+        pub fn borrow(
+            origin: OriginFor<T>,
+            collateral_asset: AssetIdOf<T>,
+            borrowing_asset: AssetIdOf<T>,
+            borrowing_amount: Balance,
+        ) -> DispatchResultWithPostInfo {
```

Comment:
Please, check that `borrowing_amount` is more than some minimal amount to prevent attack with blowing of `UserBorrowingInfo`.